torch/csrc/distributed/c10d/comm.h

```diff
@@ -13,4 +14,66 @@ void broadcast_coalesced(
     at::TensorList tensors,
     size_t buffer_size);
 
+// This class passes bucket contents tensor (for multiple replicas) to
+// DDP communication hook.
+// Optionally in the future this can be enhanced with parameter to bucket
+// mappings as well.
+class GradBucket {
+ public:
+  explicit GradBucket(std::vector<at::Tensor> tensors);
+  const std::vector<at::Tensor>& getTensors();
+
+ private:
+  std::vector<at::Tensor> tensors_;
+};
+
+// DDP's c10d reducer allows communcation hooks defined as a sub class
+// of CommHookInterface. CommHookInterface is an abstract class and can
+// be used to implement both Python and CPP hooks.
+struct TORCH_API CommHookInterface {
+ public:
+  virtual ~CommHookInterface() {}
+
+  // runHook takes a GradBucket type bucket and passses the tensors of
+  // this grad bucket to hook's callaback. This function is called once
+  // the bucket is ready. The hook can perform whatever processing is
+  // needed and return a Future that will hold the new value of the grad
+  // bucket's tensors once ready.
+  virtual c10::intrusive_ptr<torch::jit::Future> runHook(
+      const GradBucket& bucket) = 0;
+
+  // Once the grad bucket of Future is ready, c10d reducer will call this
+  // function to get the resulting tensors of the grad bucket. Then c10d
+  // reducer will use these tensors and copy grads to the grads of individual
+  // parameters.
+  virtual std::vector<at::Tensor> processFuture(c10::IValue future_value) = 0;
+};
+
+// PythonCommHook enables registering a python hook to c10d reducer and is a
+// sub class of CommHookInterface.
+class TORCH_API PythonCommHook : public CommHookInterface {
+ public:
+  // The constructor takes a state and a callable hook. Inputs are Python
+  // objects. The state is passed to the hook in runHook function can be used to
+  // maintain and update any state information that users would like to maintain
+  // as part of the training process. The hook can perform whatever processing
+  // user specified and return a Future indicating completion of any async work.
+  PythonCommHook(py::object state, py::object hook);
+
+  ~PythonCommHook() override {
+    pybind11::gil_scoped_acquire ag;
+    state_.dec_ref();
```

Comment:
any reason not setting `state_.ptr()` and `hook_.ptr()` to `nullptr`? See the code example below:

https://github.com/pytorch/pytorch/blob/7e82382ad577078cbcf5b21c623baca567461997/torch/csrc/distributed/rpc/python_rpc_handler.cpp#L61-L67